### PR TITLE
Hyperexponential coordinates

### PR DESCRIPTION
**Pull request by brryan**
## PR Summary

An alternative (going back to ~Tchekhovskoy+ 2011) to having an outer boundary near a disk is to use hyperexponential coordinates such that, beyond some break radius, zones grow radially in size very rapidly with radius.

This PR implements the hyperexponential coordinates based on how they are implemented in KHARMA.

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by calling `scripts/bash/format.sh`.
- [x] Explain what you did.
- [x] Do at least some rudimentary testing.
- [x] Make sure this resolves to the `flatgcov` treatment in #165

### CODE REVIEW

File: src/geometry/mckinney_gammie_ryan.hpp [hunk 1]
```diff
@@ -66,7 +72,10 @@ class McKinneyGammieRyan {
     C[1] = th;
     C[2] = X3;
 
-    const Real drdX1 = std::exp(X1);
+    // const Real drdX1 = std::exp(X1);
+    const Real hexp_dr = X1 - hexp_br_;
+    const Real drdX1 =
+        std::exp(X1 + (hexp_dr > 0.) * hexp_csq_ * std::pow(hexp_dr, hexp_nsq_));
```
Comment (Yurlungur) [inline src/geometry/mckinney_gammie_ryan.hpp:78]: nice use of masking. I was worried there'd be an if statement.

**Review by Yurlungur — APPROVED** [review]
We should probably make sure this doesn't break anything and also add a test. Other than that, :+1: nice improvement.

I *would* like someday to fully compactify the domain and go all the way to physical infinity at $X1=1$ or something. But that can be a problem for future us.

File: src/geometry/mckinney_gammie_ryan.hpp (hunk 1 above)
Comment (bprather) [reply src/geometry/mckinney_gammie_ryan.hpp:78]: Is drdX1 supposed to be the derivative?  If so this is missing a term, right?  (vs [here](https://github.com/AFD-Illinois/kharma/blob/feature/multizone-dev/kharma/coordinates/coordinate_systems.hpp#L476))

Incidentally, if Phoebus uses numerical first derivatives/transformation matrices, why?

File: src/geometry/mckinney_gammie_ryan.hpp [hunk 2]
```diff
@@ -191,6 +202,9 @@ class McKinneyGammieRyan {
   Real x0_ = 0; // start point of smooth region
   Real smooth_ = 0.5;
   Real norm_;
+  Real hexp_br_ = 1000.;
+  Real hexp_nsq_ = 1.;
```
Comment (bprather) [inline src/geometry/mckinney_gammie_ryan.hpp:206]: As I discovered when I tried to run things with this value, nsq=1 is dangerous (it is the default in HARMPI and what was used in Tchekhovskoy '11 afaict though).
The first derivative of the function r(X1) is discontinuous for this value, and it exacerbates the problems with passing material over the break radius.  The effect may be much less if you use numerical first derivatives, as HARMPI does.

**Review by bprather — APPROVED** [review]
Code sharing!

File: src/geometry/mckinney_gammie_ryan.hpp (hunk 1 above)
Comment (brryan) [reply src/geometry/mckinney_gammie_ryan.hpp:78]: @Yurlungur the adroit use of masking is due to @bprather :) Does this actually save time compared to the if statement? It's still evaluating a conditional right? I could be wrong, I'm just curious.

@bprather Yes you're right I dropped a term, thanks for catching that. I'll fix. I'm not sure I understand the numerical transformation matrix concern, could you elaborate?

File: src/geometry/mckinney_gammie_ryan.hpp (hunk 2 above)
Comment (brryan) [reply src/geometry/mckinney_gammie_ryan.hpp:206]: Ah ok, I just pulled out the default values from kharma (as I interpreted them). Could you remind me what your preferred values are and I can update these defaults?

**Comment by brryan** [comment]
> Code sharing!

Does copy and pasting count? Lol

File: src/geometry/mckinney_gammie_ryan.hpp (hunk 1 above)
Comment (Yurlungur) [reply src/geometry/mckinney_gammie_ryan.hpp:78]: @brryan  the masking vectorize whereas a conditional will not... although it's not clear this vectorizes anyway. (The geometry package as a while probably does not.)

Comment (bprather) [reply src/geometry/mckinney_gammie_ryan.hpp:78]: It's not a concern, just trying to gauge.  I think in this case, numerical first+second derivatives would actually be more stable, because analytically speaking the derivative of HExp can change really quickly.  Then, a numerical second derivative of the sharp first derivative causes :sparkles: problems :sparkles:

In particular, d/dx of exp(x + pow(x,n)) is exp(x + pow(x, n))*(1 + n*exp(x + pow(x, n-1)), yeah? (Or in any case involves pow(x, n-1), I'm dashing this off).  However, if n=1, pow(x,0) is constant for *all* x, i.e., including x=0.  Thus, the derivative becomes straight-up mathematically discontinuous at x=x_br.  for n~=1, it's still pretty bad from a numerical-derivative perspective, and can lead to some gnarly connection coefficients.  And even for n=2 (which starts to *really* stretch the outer edge) it seems to have a noticeable effect, stemming I think from not preserving conn^mu_mu_nu = d/dx^nu gdet

**Comment by bprather** [comment]
> > Code sharing!
> 
> Does copy and pasting count? Lol

You changed the names!

(Honestly I like your naming better, and hadn't considered adding HExp to FMKS since I'm using it so differently, so I'll probably steal it back even more literally)

File: src/geometry/mckinney_gammie_ryan.hpp (hunk 1 above)
Comment (Yurlungur) [reply src/geometry/mckinney_gammie_ryan.hpp:78]: as a general rule, Phoebus supports analytic *and* numerical derivatives. Likely when this code is run, the derivatives will be numerical if the code is run with cached geometry.

File: src/geometry/mckinney_gammie_ryan.hpp (hunk 2 above)
Comment (Yurlungur) [reply src/geometry/mckinney_gammie_ryan.hpp:206]: :grin:

Comment (brryan) [reply src/geometry/mckinney_gammie_ryan.hpp:207]: Thanks -- updated to `hexp_nsq = 4`

**Comment by brryan** [comment]
> We should probably make sure this doesn't break anything and also add a test. Other than that, 👍 nice improvement.

The original FMKS unit test was actually good for this -- the answer is unchanged in the region below the hyperexponential break radius. I also updated that test to sample a region outside the break radius.

File: scripts/python/phoedf.py [hunk 3]
```diff
@@ -75,7 +75,7 @@ def flatten_indices(mu, nu):
     self.gcov = np.zeros([self.NumBlocks, 4, 4, self.Nx3, self.Nx2, self.Nx1])
     for mu in range(4):
       for nu in range(4):
-        self.gcov[:,mu,nu,:,:,:] = self.flatgcov[:,flatten_indices(mu,nu),:,:,:]
```
Comment (brryan) [inline scripts/python/phoedf.py:78]: I'm not immediately sure why this change was necessary, I should figure this out

Comment (brryan) [reply scripts/python/phoedf.py:78]: It was due to lack of ghost zone pruning for non-proper-variable data like `gcov` after the recent changes to parthenon output. This is properly fixed in #165

**Comment by Yurlungur** [comment]
Is this ready to go in @brryan ?

**Comment by brryan** [comment]
@Yurlungur This is ready to go in, but I think it would be easiest to merge #165 first when everyone thinks thats ready, then I can deal with the merge conflicts and undoing the less-good fix to `gcov` ghost cell pruning in `phoedf.py` that is currently in this PR if thats OK with you

File: src/geometry/fmks.cpp [hunk 4]
```diff
@@ -46,6 +46,9 @@ void Initialize<FMKSMeshBlock>(ParameterInput *pin, StateDescriptor *geometry) {
   Real alpha = pin->GetOrAddReal("coordinates", "poly_alpha", 14);
   Real x0 = pin->GetReal("parthenon/mesh", "x1min");
   Real smooth = pin->GetOrAddReal("coordinates", "smooth", 0.5);
+  Real hexp_br = pin->GetOrAddReal("coordinates", "hexp_br", 1.e6);
```
Comment (brryan) [inline src/geometry/fmks.cpp:49]: Hyperexponential coordinates are on by default... I only see this value being a problem for maybe very large cutout simulations that don't expect hyperexponential coordinates. Maybe this default value should be `std::numeric_limits<Real>::max()`?

File: scripts/bash/format.sh [hunk 5]
```diff
@@ -59,6 +60,6 @@ for f in $(git grep --untracked -ail res -- :/*.py); do
     if [ ${VERBOSE} -ge 1 ]; then
        echo ${f}
     fi
-    ${PFM} ${REPO}/${f}
+    ${PFM} -q ${REPO}/${f}
```
Comment (brryan) [inline scripts/bash/format.sh:63]: Too many cake emojis for unchanged files

File: src/geometry/fmks.cpp (hunk 4 above)
Comment (Yurlungur) [reply src/geometry/fmks.cpp:49]: I would prefer `std::numeric_limits<Real>::max()` yeah

Comment (brryan) [reply src/geometry/fmks.cpp]: `numeric_limits::max()` caused problems so I just used `1.e30` with a check that that's smaller than `Real`.
